### DISCUSSION TITLE
Web Workers throws a CSP error - could we disable them?

### DISCUSSION THREAD

**Discussion by ghiscoding** [101arrowz/fflate#205] [Q&A] [answered]
Hello, I'm using fflate in a new project [Excel-Export-Vanilla](https://github.com/ghiscoding/excel-builder-vanilla) which sometime throws an error but only when the file is big enough and is starting to use Web Workers. Googling the error, I see that I might to adjust the CSP (Content Security Policy) rules, maybe something like `worker-src 'self' blob: ;`, however it would be nice if `fflate` had an option to not disable Web Workers (I know that they are used only when using `zip` and I could use `zipAsync` to disable them but that would require 2 implementations which I would prefer to avoid). 

Basically, I'd like to know if there's a different approach to this error and/or if `fflate` could have a way to disable Web Worker. I'd like to keep using `zip` but provide a way to users of my lib to disable Web Worker if they want with just an option. 

```ts
browser.js:12 Refused to create a worker from 'blob:http://localhost:8888/2667b3f9-5c8d-4510-9087-7b6cf1d24531' 
because it violates the following Content Security Policy directive: "script-src 'self' 'unsafe-inline'". 
Note that 'worker-src' was not explicitly set, so 'script-src' is used as a fallback.
...

browser.js:12 Uncaught (in promise) DOMException: Failed to construct 'Worker': Access to the script at 
'blob:http://localhost:8888/2667b3f9-5c8d-4510-9087-7b6cf1d24531' is denied by the document's Content Security Policy.
```

For confirmation, adding `worker-src 'self' blob:;` did fix the CSP error, but still an option to disable Web Worker when using `zip` would be nice.

**Comment by 101arrowz** [comment, marked as answer]
To clarify, `fflate` offers two ZIP functions: `zipSync` and `zip`. `zipSync` runs synchronously and without any Web Workers. `zip` runs asynchronously and uses Web Workers. There is no reason to run `zip` in an environment that does not support blob-based Web Workers; the only reason you'd use it instead of `zipSync` is to take advantage of multicore compression using worker threads. The way to disable Web Workers is simply to use `zipSync` instead.

> I'd like to keep using zip but provide a way to users of my lib to disable Web Worker if they want with just an option.

I recommend just conditionally using `zip` or `zipSync` based on the user options. Having two implementations is not as convoluted as it seems, and `fflate` is designed in such a way that it will only increase your bundle size by 200 bytes or so. For example:

```js
import { zip, zipSync } from 'fflate';

const zipWorker = files => {
  return new Promise((resolve, reject) => {
    zip(files, (err, result) => err ? reject(err) : resolve(result));
  });
}
const zipNoWorker = async files => {
  return zipSync(files);
}

const zipFunc = canUseWorkers ? zipWorker : zipNoWorker;

async function main() {
  const result = await zipFunc({ 'yourfile.bin': new Uint8Array(10) });
  console.log(result); // your zip file
}
```

**Reply by ghiscoding** [reply]
> The way to disable Web Workers is simply to use `zipSync` instead.

Thanks that is what I figured was the approach to use, also great to see a demo of that. I'll go with that, thanks again for the great lib, I now officially use it :)

**Comment by mkassm** [comment]
Hi @101arrowz,

Thanks for clarification, I have the same error with `unzip` method and there is no such as method for async version like `unzipAsync`. Is there any workaround for this?
